packages/react-sdk/src/components/Whiteboard/ElementBehaviors/utils.ts

```diff
@@ -271,6 +399,156 @@ function findPathElementsToDisconnect(
   return res;
 }
 
+export type ElementFrameChange =
+  | {
+      oldFrameId?: string;
+      newFrameId: string;
+    }
+  | {
+      oldFrameId: string;
+      newFrameId?: string;
+    };
+
+/**
+ * Find changes to be applied to elements regarding attaching to frame.
+ * @param elementAttachFrame element to frame attachment
+ * @param elements current elements
+ */
+export function findElementFrameChanges(
+  elementAttachFrame: Record<string, string>,
+  elements: Elements,
+): Record<string, ElementFrameChange> {
+  const elementFrameChanges: Record<string, ElementFrameChange> = {};
+
+  for (const [elementId, element] of Object.entries(elements)) {
+    if (
+      element.type !== 'frame' &&
+      element.attachedFrame !== elementAttachFrame[elementId]
+    ) {
+      elementFrameChanges[elementId] = {
+        oldFrameId: element.attachedFrame,
+        newFrameId: elementAttachFrame[elementId],
+      };
+    }
+  }
+
+  return elementFrameChanges;
+}
+
+/**
+ * Invert the changes to be based on frame.
+ * @param changeElementFrame element to frame changes
+ * @returns frame to elements changes
+ */
+export function invertChangeElementFrame(
+  changeElementFrame: Record<string, ElementFrameChange>,
+): Record<string, FrameElementsChange> {
+  const changeFrameElements: Record<string, FrameElementsChange> = {};
+
+  for (const [elementId, change] of Object.entries(changeElementFrame)) {
+    const { oldFrameId, newFrameId } = change;
+    if (oldFrameId && changeFrameElements[oldFrameId] === undefined) {
+      changeFrameElements[oldFrameId] = {
+        attachElementIds: [],
+        detachElementIds: [],
+      };
+    }
+    if (newFrameId && changeFrameElements[newFrameId] === undefined) {
+      changeFrameElements[newFrameId] = {
+        attachElementIds: [],
+        detachElementIds: [],
+      };
+    }
+    if (oldFrameId) {
+      changeFrameElements[oldFrameId].detachElementIds.push(elementId);
+    }
+    if (newFrameId) {
+      changeFrameElements[newFrameId].attachElementIds.push(elementId);
+    }
+  }
+
+  return changeFrameElements;
+}
+
+/**
+ * Take elements and frames and find the element to frame intersection via bounding rectangles.
+ * @param elements elements to check
+ * @param frameElements frame elements
+ * @returns element to frame intersection
+ */
+export function findElementAttachFrame(
+  elements: Elements,
+  frameElements: Elements<FrameElement>,
+): Record<string, string> {
+  const elementAttachFrame: Record<string, string> = {};
+
+  for (const [elementId, element] of Object.entries(elements)) {
+    if (element && element.type !== 'frame') {
+      const frameElementId = findFrameToAttach(element, frameElements);
+      if (frameElementId) {
+        elementAttachFrame[elementId] = frameElementId;
+      }
+    }
+  }
+
+  return Object.fromEntries(
+    Object.entries(elementAttachFrame).filter(([elementId, frameId]) => {
+      const element = elements[elementId];
```

Comment:
TypeScript doesn't cry that `element` could be `undefined`?

Is `"noUncheckedIndexedAccess": true` not set in tsconfig.json?

Reply:
Extra `undefined` check is added, though `undefined` could not happen in this place.